### PR TITLE
feat: impl From<Principal> for Subaccount

### PR DESCRIPTION
**Pull request by hpeebles**
# Description

This PR makes it trivial to convert a Principal into a Subaccount.
Principals are often used as Subaccounts to differentiate between funds controlled by the same Principal (eg. when creating a canister the Subaccount is the Principal of the new canister's controller).

# How Has This Been Tested?

I used this forked version to successfully top up a canister using the canister's Principal as the Subaccount.
I have also added a unit test.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [x] I have edited the CHANGELOG accordingly.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

File: library/ic-ledger-types/src/lib.rs [hunk 1]
```diff
@@ -123,6 +123,16 @@ impl fmt::Display for Tokens {
 )]
 pub struct Subaccount(pub [u8; 32]);
 
+impl From<Principal> for Subaccount {
+    fn from(principal: Principal) -> Self {
+        let mut subaccount = [0; 32];
+        let principal = principal.as_slice();
+        subaccount[0] = principal.len().try_into().unwrap();
+        subaccount[1..1 + principal.len()].copy_from_slice(principal);
+        Subaccount(subaccount)
```
Comment (ericswanson-dfinity) [inline library/ic-ledger-types/src/lib.rs:132]: How about adding a test?  There are a few at the bottom of the file.

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: library/ic-ledger-types/src/lib.rs (hunk 1 above)
Comment (hpeebles) [reply library/ic-ledger-types/src/lib.rs:132]: Yeah I considered that but I didn't think of a suitable way to test this that isn't simply duplicating the logic in the test.
There are no existing functions exposed by `Subaccount` which could be used to verify the outcome of the `Principal` to `Subaccount` conversion.

Comment (hpeebles) [reply library/ic-ledger-types/src/lib.rs:132]: I could convert a `Principal` into a `Subaccount` and then check that the bytes match what is expected.
But the only way I'd be able to get the expected bytes is by running the function itself.

Comment (hpeebles) [reply library/ic-ledger-types/src/lib.rs:132]: Ok I've found a way that works and have added a unit test.

**Review by ericswanson-dfinity — APPROVED** [review]

**Comment by hpeebles** [comment]
Who merges this? Should I be merging it?

**Pull request merged**